### DISCUSSION TITLE
[Apex] Is InterfaceNames available to be used in a custom XPath rule?

### DISCUSSION THREAD

**Discussion by thvd** [pmd/pmd#4950] [Q&A] [answered]
Context: 

Given i have an apex class implementing `Queueable` interface. Using PMD i want to ensure that always also the `Finalizer` interface is also implemented to handle exceptions, to for example write them to a log. 

The question i have is are the implemented interfaces available for a custom XPath rule? 
I have already looked in the designer and could find it, so probably the answer is no, but i am wondering why? And is this something i can contribute to?

I could find it however in the Java code: https://github.com/pmd/pmd/blob/d4b99bb047c832f4b0b7153dc91b23f6b8d19565/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTUserClass.java#L43

**Comment by jsotuyod** [comment]
No, getters returning lists are not exposed as attributes in xpath… I think you have a legit use case here, but right now I believe it's only feasible with rules written in Java…

**Reply by oowekyala** [reply]
Getters having a list type unfortunately can not be supported in XPath because of Saxon's implementation. However it is possible to write a custom XPath function to return that.

**Comment by jsotuyod** [comment, marked as answer]
For completeness, I've just submitted a PR for PMD 7.1.0 that would expose collection attributes such as `getInterfaceNames()` as an XPath attribute through a sequence. https://github.com/pmd/pmd/pull/4969

With that you should be able to write

```xml
//UserClass[
  not(empty(index-of(@InterfaceNames, 'Queueable'))). (: implements Queueable :)
  and empty(index-of(@InterfaceNames, 'Finalizer')) (: but doesn't implement Finalizer :)
]
```

**Reply by oowekyala** [reply]
Not that this `not(empty(index-of(` is not the best way to write that in xpath. You can directly write `@InterfaceNames = 'Queuable' and not(@InterfaceNames = 'Finalizable')`. The equals operator tests for equality of any two members of the left and right sequence.
